src/Microsoft.AspNetCore.Mvc.Core/ControllerAttribute.cs

```diff
@@ -0,0 +1,17 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+
+namespace Microsoft.AspNetCore.Mvc
+{
+    /// <summary>
+    /// Indicates that the type and any derived types that this attribute is applied to
+    /// are considered a controller by the default controller discovery mechanism, unless
+    /// <see cref="NonControllerAttribute"/> is applied to any type in the hierarchy.
```

Comment:
As an end user, I do agree with @dougbu, but MVC doesn't follow this rule already (e.g. `AlllowAnonymous` attribute). It would be better to not introduce `NonController` attribute at all.